service/history/api/getworkflowexecutionrawhistoryv2/api.go

```diff
@@ -0,0 +1,288 @@
+		BranchToken: targetVersionHistory.GetBranchToken(),
+		// GetWorkflowExecutionRawHistoryV2 is exclusive exclusive.
+		// ReadRawHistoryBranch is inclusive exclusive.
+		MinEventID:    pageToken.GetStartEventId() + 1,
+		MaxEventID:    pageToken.GetEndEventId(),
+		PageSize:      pageSize,
+		NextPageToken: pageToken.PersistenceToken,
+		ShardID:       shardID,
+	})
+	if err != nil {
+		if _, isNotFound := err.(*serviceerror.NotFound); isNotFound {
+			// when no events can be returned from DB, DB layer will return
+			// EntityNotExistsError, this API shall return empty response
+			return &historyservice.GetWorkflowExecutionRawHistoryV2Response{
+				Response: &adminservice.GetWorkflowExecutionRawHistoryV2Response{
+					HistoryBatches: []*commonpb.DataBlob{},
+					NextPageToken:  nil, // no further pagination
+					VersionHistory: targetVersionHistory,
+				},
+			}, nil
+		}
+		return nil, err
+	}
+
+	pageToken.PersistenceToken = rawHistoryResponse.NextPageToken
+	size := rawHistoryResponse.Size
+	shard.GetMetricsHandler().Histogram(metrics.HistorySize.GetMetricName(), metrics.HistorySize.GetMetricUnit()).Record(
+		int64(size),
+		metrics.NamespaceTag(ns.Name().String()),
+		metrics.OperationTag(metrics.AdminGetWorkflowExecutionRawHistoryV2Scope),
+	)
+
+	result :=
+		&adminservice.GetWorkflowExecutionRawHistoryV2Response{
+			HistoryBatches: rawHistoryResponse.HistoryEventBlobs,
+			VersionHistory: targetVersionHistory,
+			HistoryNodeIds: rawHistoryResponse.NodeIDs,
+		}
+	if len(pageToken.PersistenceToken) == 0 {
+		result.NextPageToken = nil
+	} else {
+		result.NextPageToken, err = api.SerializeRawHistoryToken(pageToken)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &historyservice.GetWorkflowExecutionRawHistoryV2Response{
+		Response: result,
+	}, nil
+}
+
+func validateGetWorkflowExecutionRawHistoryV2Request(
+	request *historyservice.GetWorkflowExecutionRawHistoryV2Request,
+) error {
+
+	req := request.Request
+	execution := req.Execution
+	if execution.GetWorkflowId() == "" {
+		return consts.ErrWorkflowIDNotSet
+	}
+	// TODO currently, this API is only going to be used by re-send history events
+	// to remote cluster if kafka is lossy again, in the future, this API can be used
+	// by CLI and client, then empty runID (meaning the current workflow) should be allowed
+	if execution.GetRunId() == "" || uuid.Parse(execution.GetRunId()) == nil {
+		return consts.ErrInvalidRunID
+	}
+
+	pageSize := int(req.GetMaximumPageSize())
+	if pageSize <= 0 {
+		return consts.ErrInvalidPageSize
+	}
+
+	if req.GetStartEventId() == common.EmptyEventID &&
+		req.GetStartEventVersion() == common.EmptyVersion &&
+		req.GetEndEventId() == common.EmptyEventID &&
+		req.GetEndEventVersion() == common.EmptyVersion {
+		return consts.ErrInvalidEventQueryRange
+	}
+
+	return nil
+}
+
+func SetRequestDefaultValueAndGetTargetVersionHistory(
```

Comment:
nit: un-export?

Reply:
cannot un-export because the tests are under `./service/history/`